### PR TITLE
impl(GCS+gRPC): asynchronous object reader

### PR DESCRIPTION
**Pull request by coryan**
Introduce `AsyncReader`, the public API to asynchronously read object contents.

Part of the work for #12116

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12688)

### CODE REVIEW

File: google/cloud/storage/async_reader.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+namespace google {
+namespace cloud {
+namespace storage_experimental {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * A handle for streaming downloads.
+ *
+ * Application use this object to handle asynchronous streaming downloads. The
```
Comment (devbww) [inline google/cloud/storage/async_reader.h:33]: s/Application/Applications/

File: google/cloud/storage/async_reader.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ * application repeatedly calls `Read()` until it has received all the data it
+ * wants.
+ *
+ * The destructor of this object
```
Comment (devbww) [inline google/cloud/storage/async_reader.h:37]: "The destructor of this object ..."?

File: google/cloud/storage/async_reader.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+   * an object can fail even after the download starts, thus this function
+   * returns wraps the value with `StatusOr<>`. A successful end-of-stream is
+   * indicated by an invalid token.
+   *
+   * @par Example (with coroutines)
+   * @code
+   * future<std::int64_t> CountLines(AsyncReader reader, AsyncToken token) {
+   *   std::int64_t result = 0;
+   *   while (token.valid()) {
+   *     // await for next chunk and throw on error.
```
Comment (devbww) [inline google/cloud/storage/async_reader.h:72]: s/await for/wait for/

File: google/cloud/storage/async_reader.cc [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  explicit Discard(std::unique_ptr<AsyncReaderConnection> impl)
+      : impl_(std::move(impl)) {}
+
+  void Loop() {
+    auto self = shared_from_this();
+    self->impl_->Read().then([self](auto f) mutable {
+      auto s = std::move(self);
+      auto response = f.get();
+      if (absl::holds_alternative<Status>(response)) {
+        s.reset();
```
Comment (devbww) [inline google/cloud/storage/async_reader.cc:47]: Wouldn't this happen during the `return` anyway?

File: google/cloud/storage/async_reader.h [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+  ///@{
+  /// @name Move-only.
+  AsyncReader(AsyncReader&&) noexcept = default;
+  AsyncReader& operator=(AsyncReader&&) noexcept = default;
+  AsyncReader(AsyncReader const&) = delete;
+  AsyncReader& operator=(AsyncReader const&) = delete;
+  ///@}
+
+  /// Destructor.
```
Comment (devbww) [inline google/cloud/storage/async_reader.h:56]: Perhaps whatever you were going to write above belongs here?

Comment (coryan) [reply google/cloud/storage/async_reader.h]: Fixed, thanks

File: google/cloud/storage/async_reader.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/async_reader.cc]: Done.

File: google/cloud/storage/async_reader.h (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/async_reader.h]: Done. Though I liked the `await` / `co_await` match.

File: google/cloud/storage/async_reader.h (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/async_reader.h]: Fixed.

File: google/cloud/storage/async_reader.h (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/async_reader.h]: Fixed.

File: google/cloud/storage/async_reader.h [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_ASYNC_READER_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_ASYNC_READER_H
+
+#include "google/cloud/storage/async_object_responses.h"
+#include "google/cloud/storage/async_reader_connection.h"
+#include "google/cloud/storage/async_token.h"
+#include "google/cloud/status_or.h"
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/async_reader.h:21]: Done.

**Review by devbww — APPROVED** [review]

**Pull request merged**